frontend/src/modules/dataset/dataset-browser.reducer.tsx

```diff
@@ -0,0 +1,262 @@
+/**
+  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License").
+  You may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+import { Reducer } from 'react';
+import { breadcrumbFromDataset, lastComponent } from './dataset-browser.utils';
+import { ServerRequestProps } from '../../shared/util/table-utils';
+import { DatasetContext } from '../../shared/util/dataset-utils';
+import { PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
+import { DatasetType, IDataset } from '../../shared/model/dataset.model';
+import axios from '../../shared/util/axios-utils';
+import { BreadcrumbGroupProps, PaginationProps } from '@cloudscape-design/components';
+import { DatasetBrowserManageMode } from './dataset-browser.types';
+
+/**
+ * Thunk parameter for {@link getDatasetContents}. Extends {@link ServerRequestProps} with
+ * additional properties for fetching dataset contents.
+ */
+export type DatasetServerRequestProps = ServerRequestProps & {
+    datasetContext: Partial<DatasetContext>,
+    username?: string
+};
+
+/**
+ * Base type for the contents of a dataset
+ */
+type DatasetResourceBase = {
+    type: string,
+    bucket: string,
+    name: string
+};
+
+/**
+ * Object variant of {@link DatasetResourceBase}
+ */
+export type DatasetResourceObject = DatasetResourceBase & {
+    type: 'object',
+    key: string,
+    size?: number,
+    file?: File,
```

Comment:
I'll have to look, I think `size` doesn't need to be optional anymore. `file` is optional because it is used to store the actual `File` object when uploading. There is already a related `DatasetResourcePrefix` object so I don't think another subtype just for uploading makes sense.